ES|QL: better validation for LIKE and RLIKE patterns

luigidellaquila opened the pull request:
Early validation (at parse time) of RLIKE patterns, so that we can return meaningful error messages when the regex is invalid

Fixes https://github.com/elastic/elasticsearch/issues/112485

luigidellaquila committed:
ES|QL: better validation for RLIKE patterns

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

luigidellaquila committed:
Update docs/changelog/112489.yaml

elasticsearchmachine commented:
Hi @luigidellaquila, I've created a changelog YAML for you.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/ExpressionBuilder.java:
```diff
@@ -605,7 +605,18 @@ public Expression visitRegexBooleanExpression(EsqlBaseParser.RegexBooleanExpress
+                try {
+                    // the pattern could be invalid
+                    // this forces to build the automaton, so that in case it will fail right away with a meaningful error message
+                    rlike.pattern().matchesAll();
```
Question/Thought: Maybe sometimes building the automaton could be expensive, and in rare cases even irrelevant (when RLIKE get's optimized away); maybe we can move the validation a bit later, but probably this is just fine the way it is :)

I also wonder if the memory taken up by the automata of many RLIKEs can be significant; we already know that our query plans sometimes take up quite some memory which needs to be tracked.

luigidellaquila replied:
> maybe we can move the validation a bit later, but probably this is just fine the way it is :)

I put it in the parser just because I consider it a syntax error (for the same reason I throw a ParsingException, not a ValidationException). 
I don't have a very strong opinion though, I could move it it to the Verifier, but it wouldn't make a big difference in terms of cost (the optimizations didn't happen yet). 
Doing validation in the logical optimizer seems a bit late though, and I'm not sure I'd manage to do it before pruning the possibly redundant RLIKEs.

About memory management, thanks for pointing it out. I don't have an answer unfortunately, but I'll investigate it,.

astefan replied:
Imho, I don't think the additional `Invalid regex pattern for RLIKE` message it's worth the performance penalty of **always** trying to resolve the regex pattern. @luigidellaquila before this change here where is the first place where we try to build the automaton?

luigidellaquila replied:
It's during logical plan optimization, in [ReplaceRegexMatch](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/ReplaceRegexMatch.java#L28) rule.

Most of the time the pattern will be resolved anyway, so unless the RLIKE is completely pruned, the performance will practically be the same.

I'd really like to have a ParsingException, to point to the exact position where the problem happened, but throwing it in the optimizer doesn't seem correct...

astefan replied:
> Most of the time the pattern will be resolved anyway, so unless the RLIKE is completely pruned, the performance will practically be the same.

It's not only that. After the parser `Builder`s, there is the `IndexResolver` (and its `field_caps` call and index names validation), there is the `Analyzer` and the `Verifier` (where the execution can be stopped if field names are not right or invalid data types are used, for example) where the execution can fail. And at those points the automaton has already been built and checked needlessly.

My suggestions (up for discussion):
- throw the `ParsingException` in the `ReplaceRegexMatch` rule where you catch the IAE from the automaton creation and checks for `matchesAll` **and** `exactMatch` (a pattern can fail here as well? not only from `matchesAll`). At that point you still have the line numbers and the pattern itself (`throw new ParsingException(regexMatch.source(), "Invalid regex pattern for RLIKE [{}]: [{}]", pattern.asJavaRegex(), iae.getMessage());`). The only drawback here is the apparently wrong location to throw a ParsingException from; imho, it's worth it though.
- think about the error message itself a bit more, because `org.apache.lucene.util.automaton.Operations` throws the IAE exception in other cases as well. Even though, "Invalid regex pattern for RLIKE" is generic enough I believe...

alex-spies replied:
I slightly lean towards throwing the exception a bit later (when the pattern is first used), too, as long as the error message is the same/just as useful to the user. But on my end, that's mostly paranoia from unforeseen consequences of eagerly building the automaton; I can't think of cases off the top of my head where being eager is very bad (except that we could sometimes fail a query without even building the automaton, as Andrei pointed out).

Maybe we could move the try/catch simply into `RLikePattern.createAutomaton()`? This would guarantee we remain as lazy as we currently are.

luigidellaquila replied:
Thanks for the feedback folks!
Again, I don't have a very strong opinion on this, I'm good with moving the exception check to the optimizer 👍 
`RLikePattern.createAutomaton()` is the place that throws the exception, but yeah, it would require the Source, not sure it's worth it.
Practically, a check in `ReplaceRegexMatch` should solve the problem for now, but it could be a bit fragile in the long term, eg. if we introduce another optimization rule that builds the automaton.

astefan replied:
Also, if it's too much trouble or it's more involved for `like`, create a separate issue for it and deal with it separately.

alex-spies reviewed (approved):
LGTM, thanks Luigi!

luigidellaquila committed:
Add validation for LIKE and move validation for RLIKE in the optimizer

luigidellaquila commented:
I checked `LIKE` and I added validation for that as well.
It fails as soon as we create the pattern (ie. at parse time), so the validation is much easier.

I also moved the validation for `RLIKE` into `ReplaceRegexMatch`, but tbh I don't like this solution very much:
- it delegates syntactic validation to the optimizer, that doesn't seem correct
- it's fragile: if we change or move `ReplaceRegexMatch` (eg. swapping rule order, adding new rules before it, or moving it to local planning), the low level error could occur in another rule, at local planning level or even at execution time, that could be pretty bad
- `LIKE` and `RLIKE` are validated in two different places
- the cost of building the automaton is practically the same: `ReplaceRegexMatch` rule is in the `substitutions` batch, that is executed _before_ any condition pruning (eg. `BooleanSimplification`).

This said, I'm OK with both solutions, so if there is consensus we can move forward with this one.

luigidellaquila committed:
Merge branch 'main' into esql/validate_rlike

astefan reviewed (approved):
LGTM. Thank you for doing this for `like` as well.

> I checked `LIKE` and I added validation for that as well.

👍 

> I also moved the validation for `RLIKE` into `ReplaceRegexMatch`, but tbh I don't like this solution very much:
> - it delegates syntactic validation to the optimizer, that doesn't seem correct

Imo, it's a balance between performance and 100% correctness. And in this particular case I am willing to make the trade. The downside is that we might not catch that IAE in the future and not create a nicer error message (but we have tests for that), the upside is that we are delaying the build of the automaton as much as possible (automatons can be tricky because of the very wide range of regexes that can be thrown at them; I've seen Stackoverflows thrown from them, for example)

> - it's fragile: if we change or move `ReplaceRegexMatch` (eg. swapping rule order, adding new rules before it, or moving it to local planning), the low level error could occur in another rule, at local planning level or even at execution time, that could be pretty bad

That's a very good point, but let's not forget that the change in this PR is just a better message in case of errors triggered by an user confusion when using a specific regex. If we make the changes that you mentioned (swapping rules order, moving it locally), I am hopeful that when this happens and the test you added fails, the person in charge with the change will consider also the compromise we make now about delaying the automaton build.

> - `LIKE` and `RLIKE` are validated in two different places
> - the cost of building the automaton is practically the same: `ReplaceRegexMatch` rule is in the `substitutions` batch, that is executed _before_ any condition pruning (eg. `BooleanSimplification`).

Not quite. If we build the automaton and the Verifier finds a field that is misspelled somewhere else in the query, the query will fail and we built the automaton for nothing. Imagine this query runs in SIEM every 5 minutes or so. Imagine the mappings changed for rolledover indices for whatever reason and the query used by SIEM will make many unnecessary operations for the query to eventually fail (maybe the `field_caps` call is heavy, maybe the automaton is complex enough to require a bit more memory than usual etc)

luigidellaquila commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/ReplaceRegexMatch.java:
```diff
@@ -25,7 +26,18 @@ public ReplaceRegexMatch() {
+        if (matchesAll) {
             e = new IsNotNull(e.source(), regexMatch.field());
         } else {
             String match = pattern.exactMatch();
```
> Not quite. If we build the automaton and the Verifier finds a field that is misspelled somewhere else in the query, the query will fail and we built the automaton for nothing. Imagine this query runs in SIEM every 5 minutes or so. Imagine the mappings changed for rolledover indices for whatever reason and the query used by SIEM will make many unnecessary operations for the query to eventually fail (maybe the field_caps call is heavy, maybe the automaton is complex enough to require a bit more memory than usual etc)

That's a good point @astefan, I think it justifies this approach 👍 

> exactMatch theoretically can throw an IAE as well, but I think is very less likely to do it. Catching it in matchesAll is good enough imo.

The problem happens during the creation of the automaton, and `exactMatch()` will use the same automaton already created for `matchesAll`, so I think the risk is really low.

astefan replied:
> and exactMatch() will use the same automaton already created for matchesAll, so I think the risk is really low.

yep, you are right. I missed that.
